skill-tree/src/main/java/com/RDS/skilltree/Endorsement/EndorsementServiceImpl.java

```diff
@@ -33,6 +50,45 @@ public Page<EndorsementModel> getEndorsements(PageRequest pageRequest) {
         return endorsementRepository.findAll(pageRequest);
     }
 
+    public Page<EndorsementModelFromJSON> getEndorsementsFromDummyData(PageRequest pageRequest, String skillIDString, String userIDString) throws IOException {
+
+        try {
+            List<EndorsementModelFromJSON> endorsementModelFromJSONList = readEndorsementsFromJSON();
+            List<EndorsementModelFromJSON> filteredEndorsements = filterEndorsements(endorsementModelFromJSONList, skillIDString, userIDString);
+            return createPagedEndorsements(filteredEndorsements, pageRequest);
+        } catch (IOException e) {
+            logger.error("Error reading endorsements JSON data: {}", e.getMessage());
+            throw new IOException("Error reading endorsements JSON data", e);
+        }
+    }
+
+    private List<EndorsementModelFromJSON> readEndorsementsFromJSON() throws IOException {
+        ClassPathResource resource = new ClassPathResource(dummyEndorsementDataPath);
+        return objectMapper.readValue(resource.getInputStream(), new TypeReference<List<EndorsementModelFromJSON>>() {});
+    }
+
+    private List<EndorsementModelFromJSON> filterEndorsements(List<EndorsementModelFromJSON> endorsements, String skillIDString, String userIDString) {
+        return endorsements.stream()
+                .filter(endorsement -> isMatchingSkillId(endorsement, skillIDString))
+                .filter(endorsement -> isMatchingUserId(endorsement, userIDString))
+                .collect(Collectors.toList());
+    }
+
+    private boolean isMatchingSkillId(EndorsementModelFromJSON endorsement, String skillIDString) {
+        return skillIDString == null || skillIDString.isEmpty() || endorsement.getSkillId().equals(UUID.fromString(skillIDString));
+    }
```

Comment:
## Null Check Order:

The method checks if skillIDString is null after comparing it with endorsement.getSkillId().
If skillIDString is null, attempting to call equals on null (through endorsement.getSkillId().equals(skillIDString)) would throw a NullPointerException.

## Handling Empty String:

The check for an empty string (skillIDString.isEmpty()) might not be necessary depending on the intended behavior.
If an empty string should never be considered a match, you can remove this check.
If empty strings are allowed as matches, the current logic works as intended.

Reply:
Thanks for the detailed breakdown. These are very helpful. Will need to think some more about the above 2 comments and see if we will need modifications to handle these empty cases in a better way and get back to you

Reply:
**Null Check Order:**
For null check order since I am checking for null before the other checks, it will short circuit in case of null , let me know if there is some edge case I am missing 

**Handling Empty String:**
So , did some analysis and this  "If empty strings are allowed as matches, the current logic works as intended" is correct as for empty or null values we just treat it as nothing was passed and hence that works